### PR TITLE
wsg eth test finished

### PR DESCRIPTION
**Pull request by devlancer412**

### CODE REVIEW

**Review by chimera-defi — CHANGES_REQUESTED** [review]
can we plz add checks for price per share, and see that is behaves with new yield coming in + new deposits coming in/total supply increasing

File: test/v2/core/wsgETH.spec.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          {name: "nonce", type: "uint256"},
+          {name: "deadline", type: "uint256"},
+        ],
+      },
+      approveData,
+    );
+
+    const {r, s, v} = ethers.utils.splitSignature(signature);
+
+    // await expect(wsgEth.connect(alice).depositWithSignature(parseEther("1"), alice.address, deadline, false, v, r, s))
```
Comment (chimera-defi) [inline test/v2/core/wsgETH.spec.js]: does this not work as expected?

**Review by chimera-defi — CHANGES_REQUESTED** [review]

File: test/v2/core/wsgETH.spec.js (hunk 1 above)
Comment (devlancer412) [reply test/v2/core/wsgETH.spec.js]: Yeah. You can check this issue.
https://github.com/chimera-defi/SharedDeposit/issues/19

**Comment by devlancer412** [comment]
> can we plz add checks for price per share, and see that is behaves with new yield coming in + new deposits coming in/total supply increasing

I think this rate increasing by sgEth and wsgEth amount. but just deposit and withdraw keeps this rate to 1. How can I change this rate when test?

**Comment by chimera-defi** [comment]
@devlancer412 yep! 
so the way it works is: 
1. user mints sgETH w/ eth. At the rate 1:1
2. user mints/wraps to wsgETH w/ sgETH at the curRate
3. Rewards generated via rewardsRecvr are redirected, sgETH is minted, but is then sent to the wsgeth contract. 
4. When rewards are sent, the redemption rate for wsgETH goes up
5. user can now withdraw -> wsgETH to sgETH , at a better rate, receiving more than 1 sgETH
6. User can now redeem his >1 sgETH for equivalent eth at 1:1 

So for test, you could just admin mint some sgETH and send it to the wsgeth contract to test. 

See it in action in the rewards receiver here for e2e: https://github.com/chimera-defi/SharedDeposit/blob/main/contracts/v2/core/RewardsReceiver.sol#L32

**Comment by devlancer412** [comment]
So I need to deposit manually eth to RewardsReceiver contract to test sgEth/wsgEth rate change?

**Comment by chimera-defi** [comment]
@devlancer412 yep, exactly

**Comment by devlancer412** [comment]
https://github.com/chimera-defi/SharedDeposit/issues/23

I couldn't test that because this error

**Comment by devlancer412** [comment]
I finished test by avoiding top error.

**Comment by devlancer412** [comment]
@chimera-defi

File: test/v2/core/wsgETH.spec.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    ]);
+    await rewardsReceiver.deployed();
+
+    await sgEth.addMinter(minter.address);
+
+    // approve sgEth to alice
+    await sgEth.approve(wsgEth.address, parseEther("2"));
+    await wsgEth.deposit(parseEther("2"), alice.address);
+
+    console.log(formatEther(await wsgEth.pricePerShare()));
```
Comment (chimera-defi) [inline test/v2/core/wsgETH.spec.js]: can we remove the left over console log

File: test/v2/core/wsgETH.spec.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      .to.be.emit(wsgEth, "Withdraw")
+      .withArgs(alice.address, alice.address, alice.address, parseEther("0.5"), parseEther("0.5"));
+
+    // deposit to rewardReceiver to simulate reward
+    await deployer.sendTransaction({
+      to: rewardsReceiver.address,
+      value: parseEther("1"),
+    });
+    // sends 60% of sgEth to WSGEth contract - so current rate is 1.5/2.1
+    console.log(await rewardsReceiver.state());
```
Comment (chimera-defi) [inline test/v2/core/wsgETH.spec.js]: missing log

File: test/v2/core/wsgETH.spec.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    // increase time by reward cycle
+    await time.increase(24 * 60 * 60);
+    await wsgEth.syncRewards();
+
+    await time.increase(24 * 60 * 60);
+
+    // redeem will get 1.1 sgEth
+    await expect(wsgEth.connect(alice).redeem(parseEther("0.5"), alice.address, alice.address))
+      .to.be.emit(wsgEth, "Withdraw")
+      .withArgs(alice.address, alice.address, alice.address, parseEther("0.7"), parseEther("0.5"));
```
Comment (chimera-defi) [inline test/v2/core/wsgETH.spec.js]: are we checking the invariant? i.e. alice has the expected balance after withdraw? 
im not sure how the 10% bonus is represented here, 0.5 -> 0.55 would be expected based on the comment

also are you sure you need to manually call syncRewards? 
iirc it should be triggered on redeem.

**Review by chimera-defi — CHANGES_REQUESTED** [review]

File: test/v2/core/wsgETH.spec.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+      value: parseEther("1"),
+    });
+    // sends 60% of sgEth to WSGEth contract - so current rate is 1.5/2.1
+    console.log(await rewardsReceiver.state());
+    await rewardsReceiver.work();
+
+    await expect(wsgEth.syncRewards()).to.be.revertedWith("SyncError()");
+    // increase time by reward cycle
+    await time.increase(24 * 60 * 60);
+    await wsgEth.syncRewards();
```
Comment (chimera-defi) [inline test/v2/core/wsgETH.spec.js]: is this call necessary? it should be triggered when redeem/withdraw is called automatically. and its a error condition if it needs to be manually triggered instead.

Comment (devlancer412) [reply test/v2/core/wsgETH.spec.js]: actually syncRewards calls when user call another functions such as deposit and withdraw/redeem.
so actually not sure this function is called on time. But I should call this function to test increment of reward

File: test/v2/core/wsgETH.spec.js (hunk 4 above)
Comment (devlancer412) [reply test/v2/core/wsgETH.spec.js]: with my analysis, reward of n' epoch distributes to next round by linearly. To test easily, I should call this once.

File: test/v2/core/wsgETH.spec.js (hunk 5 above)
Comment (chimera-defi) [reply test/v2/core/wsgETH.spec.js]: it should work without the direct call. when the user calls another function 
or try to call  the following again after sync rewards: 
```
    // deposit to rewardReceiver to simulate reward
    await deployer.sendTransaction({
      to: rewardsReceiver.address,
      value: parseEther("1"),
    });
    // sends 60% of sgEth to WSGEth contract - so current rate is 1.5/2.1
    await rewardsReceiver.work();
```

File: test/v2/core/wsgETH.spec.js (hunk 4 above)
Comment (chimera-defi) [reply test/v2/core/wsgETH.spec.js]: ah cool yea updated comment looks good

**Comment by devlancer412** [comment]
@chimera-defi  Please check last updates

**Pull request closed without merging**